### PR TITLE
[Swift] Modify strings concept exercise

### PR DESCRIPTION
**Pull request by wneumann**
Move the old strings concept exercise to a string-components exercise as discussed in PR https://github.com/exercism/v3/pull/2022 as a way of breaking prerequisite cycles.

### CODE REVIEW

**Comment by wneumann** [comment]
the verify links check is currently failing because I moved the `strings` exercise to `string-concepts` so it is no-longer finding the link to strings/.docs/instructions.md in the main references/README.md and that file is not supposed to be modified manually…

File: languages/swift/reference/implementing-a-concept-exercise.md [hunk 1]
```diff
@@ -90,6 +90,10 @@ For more information, please read [this in-depth description][stub-file], [watch
 
 For more information, please read [this in-depth description][tests-file], [watch this video][video-tests-file] and check [this example tests file][example-tests-file].
 
+### Remove `README.md` file
+
+This file will be generated by the Swift Package Manager, but it needs to be removed as it conflicts with a README.md file that will be generated for the student when they attempt the exercise.
```
Comment (ErikSchierboom) [inline languages/swift/reference/implementing-a-concept-exercise.md:95]: Great addition! 👍

File: languages/swift/exercises/concept/string-components/.docs/hints.md [hunk 2]
```diff
@@ -24,8 +24,19 @@
...
 - String interpolation or concatenation can be used to glue two strings together.
 
+## 7. Get the i<sup>th</sup> letter of a sentence
+
+- Swift has a [method][string-index-offset] for computing an index based that is offset from another index.
+- The `startIndex` property of a String returns an index to the first character of the string.
+
+## 8. Shout the answer
+
+- - Swift has a [property][nsstring-uppercased-docs] for returning a string formatted into all uppercase letters.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/hints.md:34]:
```suggestion
- Swift has a [property][nsstring-uppercased-docs] for returning a string formatted into all uppercase letters.
```

File: languages/swift/exercises/concept/string-components/.docs/hints.md [hunk 3]
```diff
@@ -24,8 +24,19 @@
 
 - String interpolation or concatenation can be used to glue two strings together.
 
+## 7. Get the i<sup>th</sup> letter of a sentence
+
+- Swift has a [method][string-index-offset] for computing an index based that is offset from another index.
+- The `startIndex` property of a String returns an index to the first character of the string.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/hints.md:30]: Maybe rephrase slightly and link to an external resource to not directly give away the property to use?

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+lastLetter("Stands so high")
+// => "h"
+```
+
+## 6. Be polite
+
+Implement the function `backDoorPassword(_:)` that takes a string as input and formats it in the polite manner required for the backdoor password:
+
+```swift
+backDoorPassword('horse')
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:131]: Both single- and double quotes are used to define strings. It might for the student to see one consistent format?

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+Impatiently waits for
+Rider with harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the i<sup>th</sup> letters of each line, i.e. ["S", "u", "p", "e", "r"].
+
+The guard will then give you the word formed by the array of letters you replied with for you to put into uppercased word form and append `"!"`. Finally the password you return is `SUPER!`, and you'll get in.
+
+## Tasks
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:76]: This can be removed

```suggestion
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+```text
+Stands so high
+Huge hooves too
+Impatiently waits for
+Rider with harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the i<sup>th</sup> letters of each line, i.e. ["S", "u", "p", "e", "r"].
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:72]:
```suggestion
When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the i<sup>th</sup> letters of each line, i.e. `["S", "u", "p", "e", "r"]`.
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+```text
+Stands so high
+Huge hooves too
+Impatiently waits for
+Reins and harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the first letters of each line, i.e. ["h", "o", "r", "s", "e"].
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:48]:
```suggestion
When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the first letters of each line, i.e. `["h", "o", "r", "s", "e"]`.
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+Stands so high
+Huge hooves too
+Impatiently waits for
+Rider with harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the i<sup>th</sup> letters of each line, i.e. ["S", "u", "p", "e", "r"].
+
+The guard will then give you the word formed by the array of letters you replied with for you to put into uppercased word form and append `"!"`. Finally the password you return is `SUPER!`, and you'll get in.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:74]:
```suggestion
The guard will then give you the word formed by the array of letters you replied with for you to put into uppercased word form and append `"!"`. Finally the password you return is `"SUPER!"`, and you'll get in.
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+```text
+Stands so high
+Huge hooves too
+Impatiently waits for
+Reins and harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines and respond with the first letters of each line, i.e. ["S", "H", "I", "R", "E"].
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:22]:
```suggestion
When the guard recites the poem, you will split it into individual lines and respond with the first letters of each line, i.e. `["S", "H", "I", "R", "E"]`.
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+The guard will then give you the word formed by the array of letters you replied with for you to put into capitalized word form. Finally the password you return is `Shire`, and you'll get in.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:24]:
```suggestion
The guard will then give you the word formed by the array of letters you replied with for you to put into capitalized word form. Finally the password you return is `"Shire"`, and you'll get in.
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+## Back door
+
+In the back of the club you'll find the most renowned poets, which is like the VIP area. Because this is not for everyone, the back door process is a bit more convoluted.
+
+1. The guard will recite a poem;
+   - Again, you will have to split the poem into lines and respond with the appropriate letter _but
+     there are sometimes spaces after each sentence that will need to be removed first_.
+2. The guard will tell you all the letters you've responded with at once:
+   - You need to format the letters as a capitalized word
+   - and ask nicely, by appending `, please`
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:35]:
```suggestion
   - and ask nicely, by appending `", please"`
```

File: languages/swift/exercises/concept/string-components/.docs/instructions.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+Stands so high
+Huge hooves too
+Impatiently waits for
+Reins and harness
+Eager to leave
+```
+
+When the guard recites the poem, you will split it into individual lines, strip off any trailing spaces, and respond with the first letters of each line, i.e. ["h", "o", "r", "s", "e"].
+
+The guard will then give you the word formed by the array of letters you replied with for you to put into capitalized word form and append `", please."`. Finally the password you return is `Horse, please.`, and you'll get in.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/instructions.md:50]:
```suggestion
The guard will then give you the word formed by the array of letters you replied with for you to put into capitalized word form and append `", please."`. Finally the password you return is `"Horse, please."`, and you'll get in.
```

File: languages/swift/exercises/concept/string-components/.docs/introduction.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
+We've seen how to build up strings in previous exercises. Here we will see some of the ways to break strings apart.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/introduction.md:1]: I think the introduction might be trying to do a bit too much explaining. Ideally, the introduction should be relatively short and concise. Quoting [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd):

> Only information that is needed to understand the fundamentals of the concept and solve the exercise should be provided. Extra information should be left for the after.md.

This introduction also has a lot of code samples, but I'm unsure if they should be part of the introduction.md as to me it looks like they don't introduce new syntax. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd):

> Code examples should only be used to introduce new syntax 

Of course, the after.md can include these examples.

Maybe you could remove the code examples that don't introduce new syntax and try to shorten the introduction text a bit?

**Review by ErikSchierboom — APPROVED** [review]

File: languages/swift/config.json [hunk 14]
```diff
@@ -55,10 +55,10 @@
         "prerequisites": ["numbers", "conditionals"]
       },
       {
-        "slug": "strings",
+        "slug": "string-components",
         "uuid": "082d7aeb-ee25-4217-bd6f-82532a65797a",
-        "concepts": ["strings"],
-        "prerequisites": ["arrays"]
+        "concepts": ["string-componentss"],
```
Comment (ErikSchierboom) [inline languages/swift/config.json:60]:
```suggestion
        "concepts": ["string-components"],
```

Comment (ErikSchierboom) [reply languages/swift/config.json:60]: Should there also be a concept here regarding the indexing into a string?

File: languages/swift/exercises/concept/string-components/.docs/instructions.md (hunk 4 above)
Comment (wneumann) [reply languages/swift/exercises/concept/string-components/.docs/instructions.md:131]: Weird. No idea how that crept in there, must've been a copy-paste from the exercise I cloned the original from. Swift doesn't even do anything with single quotes at the moment.

File: languages/swift/config.json (hunk 14 above)
Comment (wneumann) [reply languages/swift/config.json:60]: Yes, I think it is important enough to make a separate concept. I also added an `opaque-indices` concept as those get introduced via string indices, which are opaque.

**Comment by wneumann** [comment]
Pushed an update to address all of the comments above.

File: languages/swift/exercises/concept/string-components/.docs/introduction.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+### Breaking strings apart
+
+In addition to building up strings out of components, strings may also be taken apart and broken into individual components. This is useful for tasks such as breaking apart a list of comma separated values into an array of the individual entities. Note that not all of these functions are in Swift's standard library and will require the importing of additional libraries. The most commonly used library is `Foundation`.
+
+```swift
+import Foundation
+
+let csv = "apple,pear,peach,orange,cherry,lime,goosberry"
+let fruit = csv.components(separatedBy: ",")
+// => ["apple", "pear", "peach", "orange", "cherry", "lime", "goosberry"]
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/introduction.md:12]: Is the `import` syntax introduced in this syntax? If so, it might be worth mentioning explicitly in the text that the `import` keyword is used to import functionality. This would also give this code example the unique bit of syntax for it to warrant inclusion in this document.

File: languages/swift/exercises/concept/string-components/.docs/introduction.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+```swift
+let index = csv.index(csv.startIndex, offsetBy: 21)
+csv[index]
+// => "g"
+```
+
+The most basic indices associated with strings are the start and end indices which point to the first character and the _position after_ the last character in a string. These indices are always guaranteed to be valid indices, but they may not necessarily point to valid characters in a string. For example, since `endIndex` is the index that points to the position _after_ the last character, trying to subscript the string at that position will raise an error.
+
+Note, however, that if the offset is not a valid index, i.e. if it would return the index before `startIndex` or after `endIndex` the operation will raise an error, crashing the program. To prevent this problem, one can specify a limiting index. This returns an optional index and it will return nil for otherwise invalid indices.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/string-components/.docs/introduction.md:31]:
```suggestion
Note, however, that if the offset is not a valid index, i.e. if it would return the index before `startIndex` or after `endIndex` the operation will raise an error, crashing the program. To prevent this problem, one can specify a limiting index. This returns an optional index and it will return `nil` for otherwise invalid indices.
```

File: languages/swift/exercises/concept/string-components/.docs/introduction.md (hunk 13 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/string-components/.docs/introduction.md:1]: I like the new version!

**Review by ErikSchierboom — APPROVED** [review]
Some minor comments.

**Comment by wneumann** [comment]
I added a little bit about importing modules, per the comments. Note, however, that this PR is currently unable to be merged as the check for dead links check fails. This is because I move the `strings` directory to `string-components` so it is no longer finding the link to strings/.docs/instructions.md in the main reference/README.md and that file is not supposed to be modified manually…

@ErikSchierboom, are you able to fix that issue?

File: languages/swift/config.json [hunk 17]
```diff
@@ -55,10 +55,10 @@
         "prerequisites": ["numbers", "conditionals"]
       },
       {
-        "slug": "strings",
+        "slug": "string-components",
         "uuid": "082d7aeb-ee25-4217-bd6f-82532a65797a",
-        "concepts": ["strings"],
-        "prerequisites": ["arrays"]
+        "concepts": ["opaque-indices", "string-components", "string-indexing"],
```
Comment (ErikSchierboom) [inline languages/swift/config.json:60]: My last suggestion is to add the importing concept here too.

**Review by ErikSchierboom — APPROVED** [review]

File: languages/swift/config.json (hunk 17 above)
Comment (wneumann) [reply languages/swift/config.json]: Importing concept added.

Comment (ErikSchierboom) [reply languages/swift/config.json]: Nice

Comment (wneumann) [reply languages/swift/config.json]: Are you able to take care of the dead link issue described above?

Comment (ErikSchierboom) [reply languages/swift/config.json]: @wneumann For renames, I'd just edit that file (it will be overwritten after merging).
